### PR TITLE
:art: Added pre orchestration tables validation

### PR DESCRIPTION
**Pull request by guderkar**

### CODE REVIEW

File: src/odap/common/tables.py [hunk 1]
```diff
@@ -1,29 +1,45 @@
...
 
     if db_name not in databases:
         return False
 
+    # pylint: disable=use-implicit-booleaness-not-comparison
     return spark.sql(f'SHOW TABLES IN {db_name} LIKE "{table_name}"').collect() != []
 
 
 def get_existing_table(table_name: str) -> Optional[DataFrame]:
     spark = SparkSession.getActiveSession()  # pylint: disable=W0641
```
Comment (ds-lukaslangr) [inline src/odap/common/tables.py:23]: change disable to error name

File: src/odap/feature_factory/feature_store.py [hunk 2]
```diff
@@ -45,3 +45,14 @@ def write_df_to_feature_store(
...
+
+
+def feature_store_table_exists(full_table_name: str) -> bool:
+    feature_store_client = FeatureStoreClient()
+
+    try:
+        feature_store_client.get_table(full_table_name)
+        return True
+
+    except Exception:  # noqa pylint: disable=broad-except
```
Comment (ds-lukaslangr) [inline src/odap/feature_factory/feature_store.py:57]: can we catch the correct Exception?

File: src/odap/feature_factory/tables_validator.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def validate_metadata_table(config: Config):
+    metadata_table = get_metadata_table(config)
+    metadata_path = get_metadata_table_path(config)
+
+    validate_hive_and_path(metadata_table, metadata_path)
+
+
+def validate_feature_store_and_hive(full_table_name: str):
+    if feature_store_table_exists(full_table_name) and not hive_table_exists(full_table_name):
+        raise Exception(f"Table '{full_table_name}' exists in Databricks Feature Store but not in hive")
```
Comment (ds-lukaslangr) [inline src/odap/feature_factory/tables_validator.py:34]: maybe use FileNotFoundError?

File: src/odap/feature_factory/tables_validator.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def validate_hive_and_path(full_table_name: str, path: str):
+    if hive_table_exists(full_table_name) and not table_path_exists(path):
+        raise Exception(f"Table '{full_table_name}' exists in hive but not in path '{path}'")
```
Comment (ds-lukaslangr) [inline src/odap/feature_factory/tables_validator.py:39]: maybe use FileNotFoundError?

**Review by ds-lukaslangr — CHANGES_REQUESTED** [review]

File: src/odap/feature_factory/feature_store.py (hunk 2 above)
Comment (guderkar) [reply src/odap/feature_factory/feature_store.py:57]: There may be different exceptions depending on existence of fs table and hive table
image
image

Comment (ds-lukaslangr) [reply src/odap/feature_factory/feature_store.py:57]: we could catch the ValueError to find out that the hive table doesn't exists right?

Comment (guderkar) [reply src/odap/feature_factory/feature_store.py:57]: Yes but it's rather confusing
```
hive not exists, feature store not exists -> ValueError
hive exists, feature store not exists  -> Exception
hive not exists, feature store exists -> ValueError
```
Also in that particular file we would be using two approaches to determine if hive table exists
- catching ValueError
- using `hive_table_exists` function

Comment (guderkar) [reply src/odap/feature_factory/feature_store.py:57]: Or maybe we could just call fs.get_table and let the library raise the exception

Comment (ds-lukaslangr) [reply src/odap/feature_factory/feature_store.py:57]: sounds good

File: odap-framework-demo/_demo/_setup/init.py [hunk 5]
```diff
@@ -6,29 +6,36 @@
...
-account_db = f'{env}_odap_features_account'
-customer_db = f'{env}_odap_features_customer'
+
+offline_sdm_db = f"{env}_odap_offline_sdm_l2"
+digi_sdm_db = f"{env}_odap_digi_sdm_l2"
+account_db = f"{env}_odap_features_account"
+customer_db = f"{env}_odap_features_customer"
+segments_db = f"{env}_odap_segments"
+targets_db = f"{env}_odap_targets"
+use_cases_db = f"{env}_odap_use_cases"
```
Comment (ds-lukaslangr) [inline odap-framework-demo/_demo/_setup/init.py:18]: this db doesnt have `{write_env}` in config.yaml

File: odap-framework-demo/config.yaml [hunk 6]
```diff
@@ -6,11 +6,11 @@ parameters:
     database: "{write_env}_odap_features_{entity}"
```
Comment (ds-lukaslangr) [inline odap-framework-demo/config.yaml:6]: usecasefactory:
    table: "odap_use_cases.use_cases"
    
use case table doesn't support `{write_env}` formatting right now

File: odap-framework-demo/config.yaml [hunk 7]
```diff
@@ -26,12 +26,12 @@ parameters:
 
   segmentfactory:
     log:
-      table: "odap_segments.export_logs"
-      path: "dbfs:/odap_segments/export_logs.delta"
+      table: "{write_env}_odap_segments.export_logs"
```
Comment (ds-lukaslangr) [inline odap-framework-demo/config.yaml:29]: is this supported in the segment factory config code?

File: odap-package/src/odap/common/tables.py [hunk 8]
```diff
@@ -1,29 +1,57 @@
...
 
 
+def feature_store_table_exists(full_table_name: str) -> bool:
+    feature_store_client = FeatureStoreClient()
+
+    try:
+        feature_store_client.get_table(full_table_name)
+        return True
+
+    except:  # noqa pylint: disable=bare-except
```
Comment (ds-lukaslangr) [inline odap-package/src/odap/common/tables.py:30]: can we at least use `except Exception`?

**Review by ds-lukaslangr — CHANGES_REQUESTED** [review]

File: odap-framework-demo/config.yaml (hunk 7 above)
Comment (guderkar) [reply odap-framework-demo/config.yaml:29]: yes
image

File: odap-package/src/odap/common/tables.py (hunk 8 above)
Comment (guderkar) [reply odap-package/src/odap/common/tables.py]: done

File: odap-framework-demo/_demo/_setup/init.py (hunk 5 above)
Comment (guderkar) [reply odap-framework-demo/_demo/_setup/init.py:18]: fixed

File: odap-framework-demo/config.yaml (hunk 6 above)
Comment (guderkar) [reply odap-framework-demo/config.yaml:6]: seems it does
image

**Review by ds-lukaslangr — APPROVED** [review]

**Pull request merged**